icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -223,23 +234,57 @@ private boolean nameIsGnFirst(PersonName name) {
             return false;
         }
 
-        String localeStr = getNameLocale(name).toString();
+        // Otherwise, search the gnFirstLocales and snFirstLocales for the locale's name.
+        // For our purposes, the "locale's name" is the locale the name itself gives us (if it
+        // has one), or the locale we guess for the name (if it doesn't).
+        Locale nameLocale = name.getNameLocale();
+        if (nameLocale == null) {
+            nameLocale = getNameLocale(name);
+        }
+
+        // this is a hack to deal with certain script codes that are valid, but not the default, for their locales--
+        // to make the parent-chain lookup work right, we need to replace any of those script codes (in the name's locale)
+        // with the appropriate default script for whatever language and region we have
+        ULocale nameULocale = ULocale.forLocale(nameLocale);
+        if (NON_DEFAULT_SCRIPTS.contains(nameULocale.getScript())) {
+            ULocale.Builder builder = new ULocale.Builder();
+            builder.setLocale(nameULocale);
+            builder.setScript(null);
+            nameULocale = ULocale.addLikelySubtags(builder.build());
+        }
+
+        // now search for the locale in the gnFirstLocales and snFirstLocales lists...
+        String localeStr = nameULocale.getName();
+        String origLocaleStr = localeStr;
+        String languageCode = nameULocale.getLanguage();
+
         do {
+            // first check if the locale is in one of those lists
             if (gnFirstLocales.contains(localeStr)) {
                 return true;
             } else if (snFirstLocales.contains(localeStr)) {
                 return false;
             }
 
-            int lastUnderbarPos = localeStr.lastIndexOf("_");
-            if (lastUnderbarPos >= 0) {
-                localeStr = localeStr.substring(0, lastUnderbarPos);
-            } else {
-                localeStr = "root";
+            // if not, try again with "und" in place of the language code (this lets us use "und_CN" to match
+            // all locales with a region code of "CN" and makes sure the last thing we try is always "und", which
+            // is required to be in gnFirstLocales or snFirstLocales)
+            String undStr = localeStr.replaceAll("^" + languageCode, "und");
```

Comment:
Huh, what is "^" + languageCode doing?

Reply:
I'm calling (or am supposedly calling) an override that takes a regex, not a straight string, and I'm using the `^` to make sure I only match the language code at the beginning of the string (so I don't inadvertently mess up some other part of the locale ID that happens to match the language code).  If you can suggest a better way to do this that's still only one line long, I'm happy to change it.

Come to think of it, I guess I could do something like...

```
String undStr = "und" + localeStr.substring(languageCode.length());
```

...but I'm worried that wouldn't be as clear.  Are there better options?